lib/spot-checker.js

```diff
@@ -0,0 +1,253 @@
+/* global Zinnia */
+
+import { queryTheIndex } from './ipni-client.js'
+import { getMinerPeerId as defaultGetMinerPeerId } from './miner-info.js'
+import { Tasker } from './tasker.js'
+
+import {
+  CarBlockIterator,
+  HashMismatchError,
+  UnsupportedHashError,
+  validateBlock
+} from '../vendor/deno-deps.js'
+import { validateHttpMultiaddr } from './multiaddr.js'
+
+/** @import {RetrievalTask} from './tasker.js' */
+/** @typedef {'all' | 'block' | 'entity'} DagScope */
+/** @typedef {'bitswap' | 'graphsync' | 'http'} RetrievalProtocol */
+/** @typedef {`${number}:${number | '*'}`} EntityBytesRange */
+
+export default class SpotChecker {
+  #fetch
+  #getMinerPeerId
+  #tasker
+
+  constructor ({
+    fetch = globalThis.fetch,
+    getMinerPeerId = defaultGetMinerPeerId
+  } = {}) {
+    this.#fetch = fetch
+    this.#getMinerPeerId = getMinerPeerId
+    this.#tasker = new Tasker({ fetch: this.#fetch })
+  }
+
+  /**
+                 *
+                 * @param {object} args
+                 * @param {RetrievalTask} args.task
+                 * @param {object} args.stats
+                 * @param {number} [args.maxByteLength]
+                 * */
+  async executeSpotCheck ({ task, stats, maxByteLength }) {
+    console.log(`Calling Filecoin JSON-RPC to get PeerId of miner ${task.minerId}`)
+    try {
+      const peerId = await this.#getMinerPeerId(task.minerId)
+      console.log(`Found peer id: ${peerId}`)
+      stats.providerId = peerId
+    } catch (err) {
+      // There are three common error cases:
+      //  1. We are offline
+      //  2. The JSON RPC provider is down
+      //  3. JSON RPC errors like when Miner ID is not a known actor
+      // There isn't much we can do in the first two cases. We can notify the user that we are not
+      // performing any jobs and wait until the problem is resolved.
+      // The third case should not happen unless we made a mistake, so we want to learn about it
+      if (err.name === 'FilecoinRpcError') {
+        // TODO: report the error to Sentry
+        console.error('The error printed below was not expected, please report it on GitHub:')
+        console.error('https://github.com/filecoin-station/spark/issues/new')
+      }
+      // Abort the check, no measurement should be recorded
+      throw err
+    }
+
+    console.log(`Querying IPNI to find retrieval providers for ${task.cid}`)
+    const { indexerResult, provider } = await queryTheIndex(task.cid, stats.providerId)
+    stats.indexerResult = indexerResult
+
+    const providerFound = indexerResult === 'OK' || indexerResult === 'HTTP_NOT_ADVERTISED'
+    if (!providerFound) return
+
+    stats.protocol = provider.protocol
+    stats.providerAddress = provider.address
+
+    await this.fetchCAR({
+      protocol: provider.protocol,
+      address: provider.address,
+      cid: task.cid,
+      stats,
+      maxByteLength
+    })
+  }
+
+  /**
+                 * @param {object} args
```

Comment:
Indentation looks way off 😅

Reply:
This applies to more jsdoc comments

Reply:
Please in the future review PRs before submitting them :)

Reply:
I will continue review after the indentation has been fixed, because that makes the diff easier to read

Reply:
I ran npx standard --fix, and that’s the result 😅

By the way, what formater have you been using on Spark?

Reply:
I've reformatted everything using VSCode and then ran `npx standard --fix` again and now it seems fine. I've been using neovim for past couple of days so maybe that's the issue.

Reply:
Not sure how this got messed up 😅

Reply:
IIRC, `standard` has no opinions on whitespace in code comments. It annoys me, I hope we will find a better solution as part of https://github.com/CheckerNetwork/roadmap/issues/166